DOCS-7206 added troubleshooting nav (#36)
mulesoft/docs-service-mesh: modules/ROOT/edits/bind-api-configure-service-mesh-CRD.adoc

= Binding Your Service to an API Using a CRD
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: service mesh, microservices, binding

> When you create the binding, Anypoint Service Mesh creates a three-way association of the service in Kubernetes,the API that you created in Anypoint Platform to represent the service, and the adapter that you created for the service. 

Reviewer: add space
```suggestion
When you create the binding, Anypoint Service Mesh creates a three-way association of the service in Kubernetes, the API that you created in Anypoint Platform to represent the service, and the adapter that you created for the service. 
```

Author: Done! Thanks!

The adapter uses the unique API ID to fetch policies and apply them to your services that are running in the cluster.

> If you created the API in the previous procedure using Anypoint Platform, you use the API ID to bind the service. However, if you used the CRD to create the API, you must use the API name to bind the service.

Reviewer: ```suggestion
If you created the API using Anypoint Platform, you use the API ID to bind the service. However, if you used the CRD to create the API, you must use the API name to bind the service.
```

Author: Done! Thanks!

* Ready: The binding is now ready to be used. 
* Failed: The binding was not created successfully. An error occurred. 
* Binding: The binding is in the process of being associated with specific application.

In the following example, a binding called `customers` is created to bind the `anypoint-service-mesh-adapter` adapter with the `customers` service. This binding is created in the `nto-payments` namespace, which is the same namespace where the adapter was previously created, and is associated with the API that has the ID as 1234.

[source,text,linenums]
----
NAMESPACE      NAME        ADAPTER                         API ID  SERVICE NAME    STATUS
nto-payments   customers   anypoint-service-mesh-adapter   1234    customers       Ready
----

== Methods for Binding an API

You can bind an API to a service using Anypoint Platform or Anypoint Service Mesh, based on your preference:

* <<bind_the_service, Bind an API created using Anypoint Platform>>: If you created your API using Anypoint Platform, use this procedure.

> * <<bind_using_ASM, Bind an API created using Anypoint Service Mesh>> : If you created your API using Anypoint Service Mesh, use this procedure.

Reviewer: remove space
```suggestion
* <<bind_using_ASM, Bind an API created using Anypoint Service Mesh>>: If you created your API using Anypoint Service Mesh, use this procedure.
```

Author: Done! Thanks!

[[bind_the_service]]
== Bind an API Created Using Anypoint Platform

Before you bind the service, ensure that you complete the task prerequisite.

*Task Prerequisite*

> From the API Administration Tab of API Manager, obtain the API ID to create the binding using this method.

Reviewer: Suggest combining as illustrated in an earlier file.

Author: Done! Thanks!

To bind the service with an API, perform the following steps:

> . Go to *Anypoint Platform > API administration* and expand the API version name to display the instances.

Reviewer: not nec
"to display the instances"

Author: Kept it. Users need to know why they are performing a specific task.

. Click an instance name and copy the API ID that is displayed. 

> Ensure that you save this API ID to a text editor.

Reviewer: Do we have to tell them where or can we say...:
```suggestion
Ensure that you save this API ID somewhere that you can access readily.
```

Author: Thanks! Done!

+
image::api-id.png[]
+

> . To retrieve the service name from Kubernetes, with which you want to bind the API, run the command:

Reviewer: ```suggestion
. Retrieve from Kubernetes the name of the service with which you want to bind the API:
```

Author: Done! Thanks!

+
`kubectl get svc -n <namespace>`

> . Verify that your microservice port is named following the requirements requested by Istio: xref:https://istio.io/docs/setup/additional-setup/requirements/[].

Reviewer: ```suggestion
. Verify that your microservice port is named following Istio standards: 
xref:https://istio.io/docs/setup/additional-setup/requirements/[]
```

Author: Done! Thanks!

> . If you created your API using Anypoint Platform, use `spec.parameters.apiId`:

Reviewer: "use `spec.parameters.apiId`" how?  Do you input this syntax? Where?

Author: In the yaml file. Good catch. Added! Thanks!

+
[source,text,linenums]
----
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: <binding name> # example: orders-api-binding
  namespace: <namespace name> # example: default
  labels: 
    service-mesh.mulesoft.com/kind: adapter-binding
spec:
  instanceRef:
    name: <adapter name> # example: bookinfo-service-mesh-adapter
  parameters:
    apiId: <api ID> # example: 15722113
    service: <service name> # example: orders-service
----
+

> The binding is activated within a few seconds and the API is successfully registered to the proxy.

Reviewer: Do we need to say that the steps were successful by including this text?
I would omit it.

Author: I'm keeping it as it reinforces the users and tells them what they should be seeing on their screen. Done! Thanks!

> . Verify the binding status using the following command:

Reviewer: ```suggestion
. Verify the binding status:
```

Author: Done! Thanks!

+
`./service-mesh adapter binding list`
+

> After you bind your API, the API in API Manager is marked as Active. You are now ready to apply policies to protect your APIs.

Reviewer: ```suggestion
The API in API Manager is marked `Active` and you can now apply policies to protect your APIs.
```

Reviewer: Okay, I am going to stop torturing you now with ideas for minimalism. You can apply them when/as you see fit.
I'll just proof the rest.

Reviewer: STOP HERE TO SUBMIT

Author: Done! Thanks!
